### PR TITLE
NSFS | NC | IAM Service - Input Validation

### PR DESCRIPTION
**Pull request by shirady**
### Explain the changes
1. Add input validation before every IAM operation.
2. Rename the file of op `get_access_key_last_used` (missing prefix `iam_`).

### Issues: 
List of GAPs:
1. Move constants to the file `iam_constants.js` (mainly from `iam_utils.js`).
2. Special handle for anonymous - currently we block all IAM User API and Access keys API on it (in the future we want to allow IAM User API on it).

#### Unit Tests
Please run:
1. `npx jest test_iam_utils.test.js`
3. `npx jest test_iam_ops_input_validation.test.js`

- [X] Doc added/updated
- [X] Tests added

### CODE REVIEW

File: src/endpoint/iam/iam_constants.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/* Copyright (C) 2024 NooBaa */
+'use strict';
+// TODO - move more constants here (mainly from iam_utils)
+const DEFAULT_MAX_ITEMS = 100;
+
+// TODO - reuse
+// function name in snake case style
+const IAM_ACTIONS = {
```
Comment (guymguym) [inline src/endpoint/iam/iam_constants.js:8]: Object.freeze

File: src/endpoint/iam/ops/iam_list_users.js [hunk 2]
```diff
@@ -12,10 +13,11 @@ async function list_users(req, res) {
 
     const params = {
         marker: req.body.marker,
-        max_items: req.body.max_items,
+        max_items: req.body.max_items ?? iam_constants.DEFAULT_MAX_ITEMS,
```
Comment (guymguym) [inline src/endpoint/iam/ops/iam_list_users.js:16]: i am not sure that nullish condition is better than simple logical or. what should be done if body max items is 0?

File: src/util/string_utils.js [hunk 3]
```diff
@@ -10,6 +10,11 @@ const email_regexp = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\
 const escape_regexp = /[\\^$.*+?()[\]{}|]/g;
 const access_key_regexp = /^[a-zA-Z0-9]{20}$/;
 const secret_key_regexp = /^[a-zA-Z0-9+/]{40}$/;
+// regex for IAM service
+const AWS_IAM_PATH_REGEXP = /(\u002F)|(\u002F[\u0021-\u007E]+\u002F)/;
```
Comment (guymguym) [inline src/util/string_utils.js:14]: using unicode code numbers for ascii characters makes it harder to read. to use forward-slash in regex you can either use \/ or my preferred [/]

Comment (guymguym) [reply src/util/string_utils.js:14]: and what is this range \u0021-\u007e? if this is needed then better document it and explain what this means.

File: src/util/string_utils.js [hunk 4]
```diff
@@ -10,6 +10,11 @@ const email_regexp = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\
...
+const AWS_USERNAME_REGEXP = /[\w+=,.@-]+/;
+const AWS_LIST_MARKER = /[\u0020-\u00FF]+/;
+const AWS_ACCESS_KEY_ID = /[\w]+/;
```
Comment (guymguym) [inline src/util/string_utils.js:17]: is \w really matching the correct character set for access keys?

File: src/util/string_utils.js [hunk 5]
```diff
@@ -10,6 +10,11 @@ const email_regexp = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\
 const escape_regexp = /[\\^$.*+?()[\]{}|]/g;
 const access_key_regexp = /^[a-zA-Z0-9]{20}$/;
 const secret_key_regexp = /^[a-zA-Z0-9+/]{40}$/;
+// regex for IAM service
+const AWS_IAM_PATH_REGEXP = /(\u002F)|(\u002F[\u0021-\u007E]+\u002F)/;
+const AWS_USERNAME_REGEXP = /[\w+=,.@-]+/;
+const AWS_LIST_MARKER = /[\u0020-\u00FF]+/;
```
Comment (guymguym) [inline src/util/string_utils.js:16]: unclear. why is this the range?

**Review by guymguym — CHANGES_REQUESTED** [review]

File: src/endpoint/iam/iam_constants.js (hunk 1 above)
Comment (shirady) [reply src/endpoint/iam/iam_constants.js:8]: Thanks, added.

File: src/endpoint/iam/ops/iam_list_users.js (hunk 2 above)
Comment (shirady) [reply src/endpoint/iam/ops/iam_list_users.js:16]: I will throw an error because in the docs of [ListUsers](https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListUsers.html#API_ListUsers_RequestParameters): 

> Valid Range: Minimum value of 1. Maximum value of 1000.

I tested it with 0 and saw an empty array (also tested it with a negative value -1 and got the items).
In the code, I decide to act as it is supposed to be according to the docs, and I will throw an error on values out of the range (including 0).
Using the nullish coalescing operator operator gives exactly what I want - according to the docs:

> If you do not include this parameter, the number of items defaults to 100.

Since you raised it now I noticed that for this parameter they used "Valid Range" While in places I tested and saw errors they used "Length Constraints". I still think it is better to throw an error and enforce the values to use as in the expected range.

Reminder - currently we haven't implemented pagination yet, so the parameter is not supported yet

File: src/util/string_utils.js (hunk 3 above)
Comment (shirady) [reply src/util/string_utils.js:14]: I took the pattern for the docs, for example ([CreateUser](https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateUser.html#API_CreateUser_RequestParameters)):

> Pattern: (\u002F)|(\u002F[\u0021-\u007E]+\u002F)

Explanation of the pattern:

> a string of characters consisting of either a forward slash (/) by itself or a string that must begin and end with forward slashes. In addition, it can contain any ASCII character from the ! (\u0021) through the DEL character (\u007F), including most punctuation characters, digits, and upper and lowercased letters.

I can add the link as an inline comment, will it be easier?

I didn't understand this comment:
> using unicode code numbers for ascii characters makes it harder to read. to use forward-slash in regex you can either use / or my preferred [/]

Would you please add a suggestion?
For the forward and backward slash I'm testing it in the code itself (the pattern doesn't include it):
`const valid_aws_path = input_path.startsWith('/') && input_path.endsWith('/') && AWS_IAM_PATH_REGEXP.test(input_path);
`

File: src/util/string_utils.js (hunk 5 above)
Comment (shirady) [reply src/util/string_utils.js:16]: I took the pattern for the docs, for example ([ListUsers](https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListUsers.html#API_ListUsers_RequestParameters)):

> Pattern: [\u0020-\u00FF]+

It was hard to find something invalid in this pattern - the example I found it:

> const invalid_characters = '\uD83D\uDE0A'; //emoji is not part of the pattern

File: src/util/string_utils.js (hunk 4 above)
Comment (shirady) [reply src/util/string_utils.js:17]: I took the pattern from the docs, for example ([UpdateAccessKey](https://docs.aws.amazon.com/IAM/latest/APIReference/API_UpdateAccessKey.html#API_UpdateAccessKey_RequestParameters)):

> Pattern: [\w]+

Explanation of the pattern:
> a string of characters that can consist of any upper or lowercased letter or digit.

File: src/util/string_utils.js (hunk 5 above)
Comment (guymguym) [reply src/util/string_utils.js:16]: Perhaps this is very specific for list users, so better rename this to AWS_LIST_USERS_MARKER

File: src/util/string_utils.js (hunk 4 above)
Comment (guymguym) [reply src/util/string_utils.js:17]: in javascript \w includes also underscore

Comment (guymguym) [reply src/util/string_utils.js:17]: and we already have an access key regex defined - 
https://github.com/noobaa/noobaa-core/blob/7b0d3e3ccbcbf45365877ef619e52fa0bac5626e/src/util/string_utils.js#L11-L12

File: src/util/string_utils.js (hunk 5 above)
Comment (shirady) [reply src/util/string_utils.js:16]: It is for list users and list access keys.
So I will suggest renaming it to `AWS_IAM_LIST_MARKER`, WDYT?

File: src/util/string_utils.js (hunk 4 above)
Comment (shirady) [reply src/util/string_utils.js:17]: 1. Why underscore should be included?
I did a test in AWS IAM - I sent access key with _ and received an error `InvalidClientTokenId`:

> The security token included in the request is invalid

and then I sent with # and received an error `ValidationError`:

> The specified value for accessKeyId is invalid. It must contain only alphanumeric characters

Note: I used `aws iam get-access-key-last-used --access-key-id <>`. took a valid access key id and changed one of its characters.

2. It is not the same regex - for example, it sets the length to be exactly 20 while for an input it can be:

> Length Constraints: Minimum length of 16. Maximum length of 128.

File: src/util/string_utils.js (hunk 5 above)
Comment (guymguym) [reply src/util/string_utils.js:16]: sounds great

File: src/endpoint/iam/ops/iam_list_users.js (hunk 2 above)
Comment (guymguym) [reply src/endpoint/iam/ops/iam_list_users.js:16]: ok, you definitely understand the difference of || and ?? well but this is always subtle for many people to read unless carefully focusing on it, imho. in any case we can use it, but I am trying to minimize it.

File: src/util/string_utils.js (hunk 3 above)
Comment (guymguym) [reply src/util/string_utils.js:14]: ok nevermind, better leave it as as.

File: src/util/string_utils.js (hunk 4 above)
Comment (guymguym) [reply src/util/string_utils.js:17]: so how about we keep the pattern as is but to avoid confusion with other definitions we rename this const to something like AWS_IAM_ACCESS_KEY_INPUT_REGEXP ?

File: src/endpoint/iam/ops/iam_list_users.js (hunk 2 above)
Comment (shirady) [reply src/endpoint/iam/ops/iam_list_users.js:16]: I can add a comment in the code: "If one did not include max_items parameter (req.body.max_items value is undefined), we use a default number DEFAULT_MAX_ITEMS", but I don't like it (it explains the operator ?? and does not give an added value), let me know what you think.

File: src/util/string_utils.js (hunk 4 above)
Comment (shirady) [reply src/util/string_utils.js:17]: I will change it.
I still think this name will be confusing because later we might use this validation in the CLI.

File: src/util/string_utils.js (hunk 5 above)
Comment (shirady) [reply src/util/string_utils.js:16]: I will change it also, thanks.

File: src/util/string_utils.js (hunk 4 above)
Comment (guymguym) [reply src/util/string_utils.js]: but you have to agree that having two different regexp that are names aws_access_key is way more confusing

File: src/endpoint/iam/ops/iam_list_users.js (hunk 2 above)
Comment (guymguym) [reply src/endpoint/iam/ops/iam_list_users.js:16]: oh but what is the type of req.body.max_items? string? if so then we actually want to parse it to a number and not send once a string and once a number (when undefined)

File: src/util/string_utils.js (hunk 4 above)
Comment (shirady) [reply src/util/string_utils.js]: I agree with this, I comment on it in case we have a better name.

File: src/endpoint/iam/ops/iam_list_users.js (hunk 2 above)
Comment (shirady) [reply src/endpoint/iam/ops/iam_list_users.js:16]: It should be a number, but now I noticed that there is an issue with AWS CLI in which it doesn't pass it.

I ran: 
`s3-nc-user-1-iam iam list-users --path-prefix /abc/ --max-items 7 --debug` 
I added the `--path-prefix` flag so we can see a flag that passes (and a flag that I think doesn't pass).
(s3-nc-user-1-iam is an alias 
for `AWS_ACCESS_KEY_ID=<access-key> AWS_SECRET_ACCESS_KEY=<secret-key> aws --no-verify-ssl --endpoint-url https://localhost:7005`).

In the debug I see AWS CLI list the argument of max-items:
```bash
2024-07-07 08:43:35,751 - MainThread - awscli.clidriver - DEBUG - Arguments entered to CLI: ['--no-verify-ssl', '--endpoint-url', 'https://localhost:7005', 'iam', 'list-users', '--path-prefix', '/abc/', '--max-items', '7', '--debug']
```

In the debug I don't see the max params passing to the request's body (PathPrefix we can see):
```bash
2024-07-07 08:43:35,836 - MainThread - botocore.endpoint - DEBUG - Making request for OperationModel(name=ListUsers) with params: {'url_path': '/', 'query_string': '', 'method': 'POST', 'headers': {'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8', 'User-Agent': 'aws-cli/2.15.36 Python/3.11.9 Darwin/23.4.0 source/arm64 prompt/off command/iam.list-users'}, 'body': {'Action': 'ListUsers', 'Version': '2010-05-08', 'PathPrefix': '/abc/'}, 'url': 'https://localhost:7005/', 'context': {'client_region': 'us-east-1', 'client_config': <botocore.config.Config object at 0x1083b49d0>, 'has_streaming_input': False, 'auth_type': None}}
```

Same issue in list-access-keys:
`s3-nc-user-1-iam iam list-access-keys --max-items 1 --user-name shira-1001 --debug`

In the debug I see AWS CLI list the argument of max-items:
```bash
2024-07-07 09:03:15,120 - MainThread - awscli.clidriver - DEBUG - Arguments entered to CLI: ['--no-verify-ssl', '--endpoint-url', 'https://localhost:7005', 'iam', 'list-access-keys', '--max-items', '1', '--user-name', 'shira-1001', '--debug']
```

In the debug I don't see the max params passing to the request's body (UserName we can see):
```bash
2024-07-07 09:03:15,228 - MainThread - botocore.endpoint - DEBUG - Making request for OperationModel(name=ListAccessKeys) with params: {'url_path': '/', 'query_string': '', 'method': 'POST', 'headers': {'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8', 'User-Agent': 'aws-cli/2.15.36 Python/3.11.9 Darwin/23.4.0 source/arm64 prompt/off command/iam.list-access-keys'}, 'body': {'Action': 'ListAccessKeys', 'Version': '2010-05-08', 'UserName': 'shira-1001'}, 'url': 'https://localhost:7005/', 'context': {'client_region': 'us-east-1', 'client_config': <botocore.config.Config object at 0x106fa89d0>, 'has_streaming_input': False, 'auth_type': None}}
```

Meanwhile, I will mention the issue in the code.

Comment (guymguym) [reply src/endpoint/iam/ops/iam_list_users.js:17]: just making sure - do you have the latest aws cli? perhaps submit an issue to its github project?

**Review by guymguym — APPROVED** [review]

File: src/endpoint/iam/ops/iam_list_users.js (hunk 2 above)
Comment (shirady) [reply src/endpoint/iam/ops/iam_list_users.js:17]: I updated my `awscli` (using brew) and still face this issue.
I opened a Github issue in their repo:
https://github.com/aws/aws-cli/issues/8800
